syntax = "proto3";
import "Common/Anime.proto";
import "Common/User.proto";

// ===========================> USERRECOMMENDATIONS PROTO FILE <===========================
// This file contains the proto file for the UserRecommendations service.
// The UserRecommendations service is responsible for asking the UserRepository for user data,
// Transforming it how ever required and sending it as a response for the Controller to use.
// ========================================================================================

// Used when user wants to get a list of users that may like the same anime as him
message users_related_by_anime_Request { // -------------> REQUEST FOR GETTING USERS THAT MAY LIKE THE SAME ANIME
    repeated Anime animes_watched = 1;
    repeated Anime animes_similar = 2;
}

message users_related_by_anime_Response {
    repeated User users = 1;
}

// Used when user wants to get a list of users that talk about the same animes as him
message users_related_by_message_Request { // -------------> REQUEST FOR GETTING USERS THAT TALK ABOUT THE SAME ANIMES
    string user_name = 1; 
}

message users_related_by_message_Response {
    repeated User users = 1;
}

message users_related_by_topics_Request {
    string user_name = 1; 
}

message users_related_by_topics_Response {
    repeated User users = 1;
}

message recomended_animeList_by_topics_Request {
    repeated string topicnames_subscribed = 1;
}

message recomended_animeList_by_topics_Response {
    repeated Anime animes = 1;
}

service UserRecommendations { // --------------------------> SERVICE FOR USERRECOMMENDATIONS
    rpc GetUsersRelatedByAnime (users_related_by_anime_Request) returns (users_related_by_anime_Response);
    rpc GetUsersRelatedByMessage (users_related_by_message_Request) returns (users_related_by_message_Response);
    rpc GetUsersRelatedByTopics (users_related_by_topics_Request) returns (users_related_by_topics_Response);
    rpc GetRecomendedAnimeListByTopics (recomended_animeList_by_topics_Request) returns (recomended_animeList_by_topics_Response);
}
